pkg/util/io/io.go

```diff
@@ -20,12 +20,114 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"path/filepath"
+	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/apimachinery/registered"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
+// ioTimeout is the timeout to use for io operations
+const ioTimeout time.Duration = 5
+
+// Lstat wraps os.Lstat in a timeout.
+func Lstat(path string) (os.FileInfo, error) {
+	return internalLstatTimeout(path, ioTimeout)
+}
+
+func internalLstatTimeout(path string, timeout time.Duration) (os.FileInfo, error) {
+	var (
+		timer      = time.NewTimer(ioTimeout)
+		resChannel = make(chan os.FileInfo)
+		errChannel = make(chan error)
+	)
+
+	defer timer.Stop()
+
+	go func() {
+		fileInfo, err := os.Lstat(path)
+		if err != nil {
+			errChannel <- err
+		} else {
+			resChannel <- fileInfo
+		}
+	}()
+
+	select {
+	case <-timer.C:
+		return nil, fmt.Errorf("Lstat on %v timed out after %v", path, timeout)
```

Comment:
But we can't have code that spawns and abandons goroutines behind the
scenes - the caller will just end up retrying and retrying and retrying.
We need to find a way to either terminate the goroutine or to make it
happen from a context that is OK to hang, or make it possible to terminate
the program.  Or not do any IO to mounted volumes...


On Thu, Aug 25, 2016 at 5:07 AM, Eric Paris <notifications@github.com>
wrote:

> In pkg/util/io/io.go
> <https://github.com/kubernetes/kubernetes/pull/31379#discussion_r76230300>
> :
>
> > +	)
> > +
> > +	defer timer.Stop()
> > +
> > +	go func() {
> > +		fileInfo, err := os.Lstat(path)
> > +		if err != nil {
> > +			errChannel <- err
> > +		} else {
> > +			resChannel <- fileInfo
> > +		}
> > +	}()
> > +
> > +	select {
> > +	case <-timer.C:
> > +		return nil, fmt.Errorf("Lstat on %v timed out after %v", path, timeout)
>
> Because NFS. Create an NFS server, mount something from it,
> delete/destroy/stop the NFS server, run ls on the client. Enjoy the
> hang... Its what NFS does...
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31379/files/32202031260c40958c2f5c398984bb8a784e1583#r76230300>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMhucegDANe6V3ew_K0EA5KqIsqgks5qjYVwgaJpZM4Jsd8Y>
> .
>